File: python/merge_k_sorted_lists.py
from typing import Optional
import sys

sys.path.append("./models")
from list_node import ListNode

sys.path.append("./utils")
from linkedListConverter import convert_to_linked_list
from linkedListConverter import convert_linked_list_to_list


# NOTE: time complexity: nlogk, where k: number of lists and n: totle elements in lists
class Solution:
    def mergeKLists(self, lists: list[Optional[ListNode]]) -> Optional[ListNode]:
        if len(lists) == 0:
            return None

        while len(lists) > 1:
            mergedLists: list[ListNode] = []
            for i in range(0, len(lists), 2):
                l1 = lists[i]
                l2 = lists[i + 1] if i + 1 < len(lists) else None
                mergedLists.append(self.mergeTwoLists(l1, l2))

            lists = mergedLists

        return lists[0]

    def mergeTwoLists(
        self, list1: Optional[ListNode], list2: Optional[ListNode]
    ) -> Optional[ListNode]:
        if list1 is None:
            return list2
        if list2 is None:
            return list1

        if list1.val < list2.val:
            list1.next = self.mergeTwoLists(list1.next, list2)
            return list1
        else:
            list2.next = self.mergeTwoLists(list1, list2.next)
            return list2


def test_mergeKLists_case_1():
    # arrange
    lists: list[ListNode] = [
        convert_to_linked_list([1, 4, 5]),
        convert_to_linked_list([1, 3, 4]),
        convert_to_linked_list([2, 6]),
    ]
    expected: list[int] = [1, 1, 2, 3, 4, 4, 5, 6]

    # act
    solution = Solution()
    actual = solution.mergeKLists(lists)

    # assert
    assert convert_linked_list_to_list(actual) == expected


def test_mergeKLists_case_2():
    # arrange
    lists: list[ListNode] = []
    expected: list[int] = []

    # act
    solution = Solution()
    actual = solution.mergeKLists(lists)

    # assert
    assert convert_linked_list_to_list(actual) == expected


def test_mergeKLists_case_3():
    # arrange
    lists: list[ListNode] = [convert_linked_list_to_list([])]
    expected: list[int] = []

    # act
    solution = Solution()
    actual = solution.mergeKLists(lists)

    # assert
    assert convert_linked_list_to_list(actual) == expected

File: rust/src/code/merge_k_sorted_lists.rs
use super::models::list_node::ListNode;

struct Solution {}

impl Solution {
    pub fn merge_k_lists(mut lists: Vec<Option<Box<ListNode>>>) -> Option<Box<ListNode>> {
        if lists.len() == 0 {
            return None;
        }

        while lists.len() > 1 {
            let mut merged_lists: Vec<Option<Box<ListNode>>> = Vec::new();
            for i in (0..lists.len()).step_by(2) {
                if i + 1 < lists.len() {
                    merged_lists.push(Self::merge_two_lists(
                        lists[i].clone(),
                        lists[i + 1].clone(),
                    ));
                } else {
                    merged_lists.push(Self::merge_two_lists(lists[i].clone(), None));
                }
            }

            lists = merged_lists;
        }

        return lists[0].clone();

        // NOTE: This would cause Time Limit Exceeded
        // let mut dummy_head = ListNode::new(0);
        // let mut pointer = &mut dummy_head;

        // for i in 0..lists.len() {
        //     if lists[i].is_none() {
        //         continue;
        //     }

        //     pointer.next = Self::merge_two_lists(pointer.next.clone(), lists[i].clone());
        // }

        // return dummy_head.next;
    }

    fn merge_two_lists(
        list1: Option<Box<ListNode>>,
        list2: Option<Box<ListNode>>,
    ) -> Option<Box<ListNode>> {
        match (list1, list2) {
            (None, None) => None,
            (Some(n), None) | (None, Some(n)) => Some(n),
            (Some(l1), Some(l2)) => {
                if l1.val >= l2.val {
                    Some(Box::new(ListNode {
                        val: l2.val,
                        next: Self::merge_two_lists(Some(l1), l2.next),
                    }))
                } else {
                    Some(Box::new(ListNode {
                        val: l1.val,
                        next: Self::merge_two_lists(l1.next, Some(l2)),
                    }))
                }
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::Solution;
    use crate::code::utils::convert_to_linked_list::convert_vec_to_linked_list;

    #[test]
    fn merge_k_lists_case_1() {
        // arrange
        let lists = vec![
            convert_vec_to_linked_list(vec![1, 4, 5]),
            convert_vec_to_linked_list(vec![1, 3, 4]),
            convert_vec_to_linked_list(vec![2, 6]),
        ];
        let expected = convert_vec_to_linked_list(vec![1, 1, 2, 3, 4, 4, 5, 6]);

        // act
        let actual = Solution::merge_k_lists(lists);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn merge_k_lists_case_2() {
        // arrange
        let lists = vec![];
        let expected = convert_vec_to_linked_list(vec![]);

        // act
        let actual = Solution::merge_k_lists(lists);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn merge_k_lists_case_3() {
        // arrange
        let lists = vec![convert_vec_to_linked_list(vec![])];
        let expected = convert_vec_to_linked_list(vec![]);

        // act
        let actual = Solution::merge_k_lists(lists);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn merge_k_lists_case_4() {
        // arrange
        let lists = vec![
            convert_vec_to_linked_list(vec![1]),
            convert_vec_to_linked_list(vec![0]),
        ];
        let expected = convert_vec_to_linked_list(vec![0, 1]);

        // act
        let actual = Solution::merge_k_lists(lists);

        // assert
        assert_eq!(expected, actual);
    }
}
